### PR TITLE
Provide Additional HTTP Verbs for Web Actions

### PR DESCRIPTION
**Pull request by dubeejw**
- Include OPTIONS, and HEAD verb support

### CODE REVIEW

**Comment by rabbah** [comment]
Related to https://github.com/openwhisk/openwhisk/pull/1868 (Which removes the meta package concept).

**Comment by dubeejw** [comment]
PG1 1180.

**Comment by dubeejw** [comment]
@rabbah, please review, or assign to someone to review.

**Comment by dubeejw** [comment]
PG2 1089.

**Comment by dubeejw** [comment]
@rabbah, tests updated, and PG is approved.

**Comment by rabbah** [comment]
the changes are good but i suggest adding a test for each of options and head specifically (ie mock a response for each of these and confirm the response is as expected).

File: tests/src/whisk/core/controller/test/MetaApiTests.scala [hunk 1]
```diff
@@ -853,6 +853,33 @@ class MetaApiTests extends ControllerTestCommon with WhiskMetaApi with BeforeAnd
                     }
                 }
         }
+
+        it should s"invoke action with options verb (auth? ${creds.isDefined})" in {
+            implicit val tid = transid()
+
+            Seq(s"$systemId/default/export_c.json").
+              foreach { path =>
+                  Options(s"$testRoutePath/$path") ~> sealRoute(routes(creds)) ~> check {
```
Comment (rabbah) [inline tests/src/whisk/core/controller/test/MetaApiTests.scala]: This test should mock a response from an options action and check that the headers are set accordingly - there is on other test in this suite that checks headers as an example to follow.

We might also add a test in `WskWebActionsTests` to confirm the options response is from the action, and NOT from the top level option in the controller.

File: tests/src/whisk/core/controller/test/MetaApiTests.scala [hunk 2]
```diff
@@ -853,6 +853,33 @@ class MetaApiTests extends ControllerTestCommon with WhiskMetaApi with BeforeAnd
...
+                  }
+              }
+        }
+
+        it should s"invoke action with options head verb (auth? ${creds.isDefined})" in {
+            implicit val tid = transid()
+
+            Seq(s"$systemId/default/export_c.json").
+              foreach { path =>
+                  Head(s"$testRoutePath/$path") ~> sealRoute(routes(creds)) ~> check {
```
Comment (rabbah) [inline tests/src/whisk/core/controller/test/MetaApiTests.scala]: Same - check that headers are set by mocking a HEAD response.

**Comment by dubeejw** [comment]
PG3 315, approved.

File: tests/src/whisk/core/controller/test/MetaApiTests.scala [hunk 3]
```diff
@@ -353,7 +365,7 @@ class MetaApiTests extends ControllerTestCommon with WhiskMetaApi with BeforeAnd
                                 "pkg" -> s"$systemId/proxy".toJson,
                                 "action" -> "export_c".toJson,
                                 "content" -> metaPayload(
-                                    m.method.name.toLowerCase,
+                                    m.method.value,
```
Comment (markusthoemmes) [inline tests/src/whisk/core/controller/test/MetaApiTests.scala]: Why change this? `m.method.name.toLowercase` seems more speaking to me than to rely on `value` returning a lower cased name?

File: core/controller/src/main/scala/whisk/core/controller/Meta.scala [hunk 4]
```diff
@@ -226,7 +226,7 @@ trait WhiskMetaApi
     private val webRoutePrefix = pathPrefix(routePath / webInvokePath)
 
     /** Allowed verbs. */
-    private lazy val allowedOperations = get | delete | post | put
+    private lazy val allowedOperations = get | delete | post | put | head | options
```
Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/controller/Meta.scala]: Any particular reason we don't just allow all of them?

**Comment by dubeejw** [comment]
@rabbah, ready for merge?

File: core/controller/src/main/scala/whisk/core/controller/Meta.scala (hunk 4 above)
Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/controller/Meta.scala]: nope

Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/controller/Meta.scala]: Should we just allow all of them?

Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/controller/Meta.scala]: TRACE, PATCH ok. CONNECT?

Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/controller/Meta.scala]: Well it just limits our users. If we don't have any particular concerns, why not?

File: tests/src/test/scala/whisk/core/cli/test/WskWebActionsTests.scala [hunk 5]
```diff
@@ -114,4 +114,23 @@ class WskWebActionsTests
             authorizedResponse.statusCode shouldBe 200
             authorizedResponse.body().asString() shouldBe namespace
     }
+
+    it should "ensure that CORS header is preserved" in withAssetCleaner(wskprops) {
```
Comment (dubeejw) [inline tests/src/test/scala/whisk/core/cli/test/WskWebActionsTests.scala]: @rabbah, please let me know if this is the correct test you wanted to be included in WskWebActionTests.

Comment (rabbah) [inline tests/src/test/scala/whisk/core/cli/test/WskWebActionsTests.scala]: excellent! yes.

Comment (dubeejw) [inline tests/src/test/scala/whisk/core/cli/test/WskWebActionsTests.scala]: Test passes locally and in Travis, but failed in PG.

**Comment by dubeejw** [comment]
Added PATCH verb. I get a compilation error when adding TRACE. Might need another import?

**Comment by dubeejw** [comment]
CORS test may not be stable.

File: tests/src/test/scala/whisk/core/cli/test/WskWebActionsTests.scala [hunk 6]
```diff
@@ -114,4 +114,23 @@ class WskWebActionsTests
...
+            val name = "webaction"
+            val file = Some(TestUtils.getTestActionFilename("corsHeaderMod.js"))
+
+            assetHelper.withCleaner(wsk.action, name) {
+                (action, _) =>
+                    action.create(name, file, annotations = Map("web-export" -> true.toJson))
+            }
+
+            val host = getServiceURL()
+            val url = host + s"/api/v1/experimental/web/$namespace/default/webaction.http"
```
Comment (rabbah) [inline tests/src/test/scala/whisk/core/cli/test/WskWebActionsTests.scala]: should this be v2?

Comment (rabbah) [inline tests/src/test/scala/whisk/core/cli/test/WskWebActionsTests.scala]: will this work with v1 given the outer header response?

File: tests/src/test/scala/whisk/core/controller/test/MetaApiTests.scala [hunk 7]
```diff
@@ -254,16 +256,18 @@ class MetaApiTests extends ControllerTestCommon with WhiskMetaApi with BeforeAnd
...
+                  nonModifierMethods.foreach { m =>
+                      m(s"$testRoutePath/$path") ~> sealRoute(routes(creds)) ~> check {
+                          status should be(NotFound)
+                      }
+                  }
                 }
         }
 
-        it should s"reject unsupported http verbs (auth? ${creds.isDefined})" in {
+        ignore should s"reject unsupported http verbs (auth? ${creds.isDefined})" in {
```
Comment (rabbah) [inline tests/src/test/scala/whisk/core/controller/test/MetaApiTests.scala]: why is this ignored?

Comment (rabbah) [inline tests/src/test/scala/whisk/core/controller/test/MetaApiTests.scala]: perhaps make the Seq empty and add a comment that we allow all the verbs.

File: tests/src/test/scala/whisk/core/cli/test/WskWebActionsTests.scala (hunk 6 above)
Comment (dubeejw) [inline tests/src/test/scala/whisk/core/cli/test/WskWebActionsTests.scala]: Agreed. Will need to fix once the related PR is in.

**Comment by rabbah** [comment]
superseded by #2004.

**Pull request closed without merging**